src/crimson/osd/pg.cc

```diff
@@ -1036,45 +1085,41 @@ PG::do_osd_ops(
-      auto error_log_fut = seastar::now();
-      epoch_t epoch = get_osdmap_epoch();
-      ceph_tid_t rep_tid = shard_services.get_tid();
-      auto last_complete = peering_state.get_info().last_complete;
-      if (op_info.may_write()) {
-        // This should be executed as OrderedExclusivePhaseT so that
-        // successive ops will not reorder.
-        // TODO: https://tracker.ceph.com/issues/61651
-        error_log_fut = submit_error_log(m, op_info, obc, e, rep_tid, version);
-      }
-      return error_log_fut.then([m, e, epoch, &op_info, rep_tid, &version, last_complete, this] {
-        auto fut = seastar::now();
-        if (!peering_state.pg_has_reset_since(epoch) && op_info.may_write()) {
-          auto it = log_entry_update_waiting_on.find(rep_tid);
-          ceph_assert(it != log_entry_update_waiting_on.end());
-          auto it2 = it->second.waiting_on.find(pg_whoami);
-          ceph_assert(it2 != it->second.waiting_on.end());
-          it->second.waiting_on.erase(it2);
-          if (it->second.waiting_on.empty()) {
-            log_entry_update_waiting_on.erase(it);
-            if (version != eversion_t()) {
-              peering_state.complete_write(version, last_complete);
-            }
-          } else {
-            fut = it->second.all_committed.get_shared_future().then(
-              [this, &version, last_complete] {
-              if (version != eversion_t()) {
-                peering_state.complete_write(version, last_complete);
-              }
-              return seastar::now();
-            });
-          }
-        }
-        return fut.then([this, m, e] {
-          return log_reply(m, e);
+    [m, &op_info, obc, this]
+    (const std::error_code& e, const ceph_tid_t& rep_tid, bool record_error) {
+    logger().error("do_osd_ops_execute::failure_func {} got error: {} record_error: {}",
+                    *m, e, record_error);
+    epoch_t epoch = get_osdmap_epoch();
+    auto last_complete = peering_state.get_info().last_complete;
+    auto fut = seastar::now();
+    if (record_error && !peering_state.pg_has_reset_since(epoch) && op_info.may_write()) {
+      logger().debug("do_osd_ops_execute::failure_func finding rep_tid {}",
+                      rep_tid);
+      ceph_assert(log_entry_version.contains(rep_tid));
+      auto it = log_entry_update_waiting_on.find(rep_tid);
+      ceph_assert(it != log_entry_update_waiting_on.end());
+      auto it2 = it->second.waiting_on.find(pg_whoami);
+      ceph_assert(it2 != it->second.waiting_on.end());
+      it->second.waiting_on.erase(it2);
+      if (it->second.waiting_on.empty()) {
+        log_entry_update_waiting_on.erase(it);
+        peering_state.complete_write(log_entry_version[rep_tid], last_complete);
+        log_entry_version.erase(rep_tid);
+        logger().debug("do_osd_ops_execute::failure_func write complete,"
+                        " erasing rep_tid {}", rep_tid);
```

Comment:
This else branch seems like it should have been moved into do_osd_ops_execute as well.  I'd probably factor it into a helper (complete_error_log) and chain it after the error_func_ptr invocation.

Reply:
#54765